31_HLSLPathTracer/app_resources/hlsl/scene.hlsl

```diff
@@ -0,0 +1,111 @@
+#ifndef _NBL_HLSL_EXT_PATHTRACING_SCENE_INCLUDED_
+#define _NBL_HLSL_EXT_PATHTRACING_SCENE_INCLUDED_
+
+#include "common.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace ext
+{
+
+template<typename Light, typename BxdfNode>
+struct Scene
+{
+    using light_type = Light;
+    using bxdfnode_type = BxdfNode;
+    using this_t = Scene<Light, BxdfNode>;
+
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t maxSphereCount = 25;
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t maxTriangleCount = 12;
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t maxRectangleCount = 12;
+
+#if SPHERE_COUNT < 1
+#define SCENE_SPHERE_COUNT 1
+#else
+#define SCENE_SPHERE_COUNT SPHERE_COUNT
+#endif
+
+#if TRIANGLE_COUNT < 1
+#define SCENE_TRIANGLE_COUNT 1
+#else
+#define SCENE_TRIANGLE_COUNT TRIANGLE_COUNT
+#endif
+
+#if RECTANGLE_COUNT < 1
+#define SCENE_RECTANGLE_COUNT 1
+#else
+#define SCENE_RECTANGLE_COUNT RECTANGLE_COUNT
+#endif
+
+    Shape<PST_SPHERE> spheres[SCENE_SPHERE_COUNT];
+    Shape<PST_TRIANGLE> triangles[SCENE_TRIANGLE_COUNT];
+    Shape<PST_RECTANGLE> rectangles[SCENE_RECTANGLE_COUNT];
```

Comment:
I want to get rid of the `#defines` so make a `SceneBase` with the 6 balls which are`CONSTEXPR_STATIC_INLINE` and then each derived class only does the light (which is movable by push constant supplied matrix)